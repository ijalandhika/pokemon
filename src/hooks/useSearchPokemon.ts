import { useState } from "react";
import { useGetPokemonsQuery } from "@/redux/services/pokemonApi";

const useSearchPokemon = () => {
  const [offset, setOffet] = useState(0);

  const { isLoading, data } = useGetPokemonsQuery(offset);

  const onUpdateOffset = (link: string) => {
    const url = new URL(link);
    const nextOffset = url.searchParams.get("offset") || 0;
    setOffet(+nextOffset);
  };

  const onFetchNext = () => {
    if (data?.next) {
      onUpdateOffset(data?.next);
    }
  };

  const onFetchPrevious = () => {
    if (data?.previous) {
      onUpdateOffset(data?.previous);
    }
  };

  return {
    isLoading,
    data,
    next: !!data?.next,
    previous: !!data?.previous,
    onFetchNext,
    onFetchPrevious,
  };
};

export default useSearchPokemon;
