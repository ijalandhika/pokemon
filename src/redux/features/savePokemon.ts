import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { toast } from "react-toastify";

import type { PokeSaved } from "@/lib/types";

type Poket = {
  value: PokeSaved[];
};

const initialState = {
  value: [],
} as Poket;

export const poketMonster = createSlice({
  name: "poketMonster",
  initialState,
  reducers: {
    saveMonster: (state, action: PayloadAction<PokeSaved>) => {
      const { payload } = action;

      const existingData = state?.value?.find((e) => e.name === payload.name);

      if (!existingData) {
        const newValue = state.value;
        newValue.push(payload);
        state.value = newValue;

        toast.success("Horray!! You've a new collections", {
          position: toast.POSITION.BOTTOM_RIGHT,
        });
      } else {
        toast.error("Upssss!! You've this already!", {
          position: toast.POSITION.BOTTOM_RIGHT,
        });
      }
    },
    releaseMonster: (state, action: PayloadAction<string>) => {
      const filtered = state.value.filter((e) => e.name !== action.payload);
      state.value = filtered;

      toast.success("This pokemon has been released", {
        position: toast.POSITION.BOTTOM_RIGHT,
      });
    },
  },
});

export const { saveMonster, releaseMonster } = poketMonster.actions;
export default poketMonster.reducer;
