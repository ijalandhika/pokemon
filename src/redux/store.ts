import { configureStore } from "@reduxjs/toolkit";
import { setupListeners } from "@reduxjs/toolkit/dist/query";
import {
  reducer as apiReducer,
  reducerPath as apiReducerPath,
  middleware as apiMiddleware,
} from "@/redux/services/pokemonApi";
import poketMonsterReducer from "./features/savePokemon";

export const store = configureStore({
  reducer: {
    poketMonsterReducer,
    [apiReducerPath]: apiReducer,
  },
  devTools: process.env.NODE_ENV !== "production",
  middleware: (getDefaultMiddleware) => {
    return getDefaultMiddleware({
      serializableCheck: false,
    }).concat(apiMiddleware);
  },
});

setupListeners(store.dispatch);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
