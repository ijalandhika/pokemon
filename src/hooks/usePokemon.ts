import { useAppDispatch, useAppSelector } from "@/redux/hooks";
import { saveMonster, releaseMonster } from "@/redux/features/savePokemon";
import type { PokemonData } from "@/lib/types";

const usePokemon = (data: PokemonData) => {
  const dispatch = useAppDispatch();

  const { id, name } = data;
  const pokeIndex = ("000" + id).slice(-3);
  const pokemons = useAppSelector((state) => state.poketMonsterReducer.value);

  const isOwned = pokemons.find((e) => e.name === name);

  const onSaveMonster = () => {
    dispatch(saveMonster({ name, index: pokeIndex }));
  };

  const onReleaseMonster = () => {
    dispatch(releaseMonster(name));
  };

  return {
    onSaveMonster,
    pokeIndex,
    isOwned,
    onReleaseMonster,
  };
};

export default usePokemon;
