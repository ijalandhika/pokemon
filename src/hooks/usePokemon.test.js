import usePokemon from "./usePokemon";

import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";

import { Providers } from "../redux/provider";
import { useAppDispatch } from "../redux/hooks";
import { saveMonster } from "../redux/features/savePokemon";

const TestComponent = () => {
  const pokemonData = {
    name: "fearow",
  };
  const { isOwned } = usePokemon(pokemonData);
  /**
   *
   * button --> dispatch new pokemon
   *
   * 1. render detail pokemon but not owned --> isOwned = false
   *
   * 2. fire event --> dispatch new pokemon
   *  --> isOwned = true
   *
   */

  const dispatch = useAppDispatch();

  const onClickButton = (e) => {
    e.preventDefault();
    dispatch(saveMonster({ name: "fearow", index: "0001" }));
  };

  return (
    <>
      <p>{isOwned ? "owned" : "not owned"}</p>
      <button name="addnew" onClick={onClickButton}>
        Add new pokemons
      </button>
    </>
  );
};

describe("useHooks", () => {
  let cont;
  beforeEach(() => {
    const { container } = render(
      <Providers>
        <TestComponent />
      </Providers>
    );
    cont = container;
  });

  it("will not owned", () => {
    expect(screen.getByText(/not owned/i)).toBeInTheDocument();
    expect(cont).toMatchSnapshot();
  });

  it("will owned the pokemon after catch", () => {
    const catchButton = screen.getByRole("button", {
      name: "Add new pokemons",
    });

    fireEvent(
      catchButton,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText(/owned/i)).toBeInTheDocument();
  });
});
