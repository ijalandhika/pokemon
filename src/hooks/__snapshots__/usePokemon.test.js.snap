// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useHooks will not owned 1`] = `
<div>
  <p>
    not owned
  </p>
  <button
    name="addnew"
  >
    Add new pokemons
  </button>
</div>
`;
